File: client/web-client-react/src/jsMain/kotlin/ro/jf/funds/client/web/api/TransactionApi.kt
package ro.jf.funds.client.web.api

import com.benasher44.uuid.uuidFrom
import com.ionspin.kotlin.bignum.decimal.BigDecimal
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.promise
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import ro.jf.funds.client.sdk.TransactionClient
import ro.jf.funds.client.web.model.JsCreateRecord
import ro.jf.funds.client.web.model.JsTransaction
import ro.jf.funds.fund.api.model.*
import ro.jf.funds.platform.api.model.Currency
import ro.jf.funds.platform.api.model.Instrument
import ro.jf.funds.platform.api.model.Label
import kotlin.js.JsExport
import kotlin.js.Promise

@JsExport
object TransactionApi {
    private val config = js("window.FUNDS_CONFIG")
    private val fundServiceUrl: String = config?.fundServiceUrl as? String ?: "http://localhost:5253"
    private val transactionClient = TransactionClient(baseUrl = fundServiceUrl)

    fun getTransaction(
        userId: String,
        transactionId: String,
    ): Promise<JsTransaction> = GlobalScope.promise {
        val transaction = transactionClient.getTransaction(uuidFrom(userId), uuidFrom(transactionId))
        JsTransaction(transaction)
    }

    fun listTransactions(
        userId: String,
        fromDate: String?,
        toDate: String?,
        fundId: String?,
        accountId: String?,
    ): Promise<Array<JsTransaction>> = GlobalScope.promise {
        val uuid = uuidFrom(userId)
        val filter = TransactionFilterTO(
            fromDate = fromDate?.takeIf { it.isNotBlank() }?.let { LocalDate.parse(it) },
            toDate = toDate?.takeIf { it.isNotBlank() }?.let { LocalDate.parse(it) },
            fundId = fundId?.takeIf { it.isNotBlank() }?.let { uuidFrom(it) },
            accountId = accountId?.takeIf { it.isNotBlank() }?.let { uuidFrom(it) },
        )
        val transactions = transactionClient.listTransactions(uuid, filter)
        transactions.map { JsTransaction(it) }.toTypedArray()
    }

    fun createTransaction(
        userId: String,
        type: String,
        dateTime: String,
        externalId: String,
        records: Array<JsCreateRecord>,
    ): Promise<JsTransaction> = GlobalScope.promise {
        val request = buildCreateTransactionTO(type, dateTime, externalId, records.toList())
        val transaction = transactionClient.createTransaction(uuidFrom(userId), request)
        JsTransaction(transaction)
    }

    fun deleteTransaction(userId: String, transactionId: String): Promise<Unit> = GlobalScope.promise {
        transactionClient.deleteTransaction(uuidFrom(userId), uuidFrom(transactionId))
    }

    private fun buildCreateTransactionTO(
        type: String,
        dateTime: String,
        externalId: String,
        records: List<JsCreateRecord>,
    ): CreateTransactionTO {
        val byRole = records.associateBy { it.role }
        val parsedDateTime = LocalDateTime.parse(dateTime)
        return when (TransactionType.valueOf(type)) {
            TransactionType.SINGLE_RECORD -> CreateTransactionTO.SingleRecord(
                dateTime = parsedDateTime,
                externalId = externalId,
                record = byRole.getValue("record").toCurrencyRecord(),
            )

            TransactionType.TRANSFER -> CreateTransactionTO.Transfer(
                dateTime = parsedDateTime,
                externalId = externalId,
                sourceRecord = byRole.getValue("sourceRecord").toCurrencyRecord(),
                destinationRecord = byRole.getValue("destinationRecord").toCurrencyRecord(),
            )

            TransactionType.EXCHANGE -> CreateTransactionTO.Exchange(
                dateTime = parsedDateTime,
                externalId = externalId,
                sourceRecord = byRole.getValue("sourceRecord").toCurrencyRecord(),
                destinationRecord = byRole.getValue("destinationRecord").toCurrencyRecord(),
                feeRecord = byRole["feeRecord"]?.toCurrencyRecord(),
            )

            TransactionType.OPEN_POSITION -> CreateTransactionTO.OpenPosition(
                dateTime = parsedDateTime,
                externalId = externalId,
                currencyRecord = byRole.getValue("currencyRecord").toCurrencyRecord(),
                instrumentRecord = byRole.getValue("instrumentRecord").toInstrumentRecord(),
            )

            TransactionType.CLOSE_POSITION -> CreateTransactionTO.ClosePosition(
                dateTime = parsedDateTime,
                externalId = externalId,
                currencyRecord = byRole.getValue("currencyRecord").toCurrencyRecord(),
                instrumentRecord = byRole.getValue("instrumentRecord").toInstrumentRecord(),
            )
        }
    }


    private fun JsCreateRecord.toCurrencyRecord(): CreateTransactionRecordTO.CurrencyRecord =
        CreateTransactionRecordTO.CurrencyRecord(
            accountId = uuidFrom(accountId),
            fundId = uuidFrom(fundId),
            amount = BigDecimal.parseString(amount),
            unit = Currency(unitValue),
            labels = labels.map(::Label),
        )

    private fun JsCreateRecord.toInstrumentRecord(): CreateTransactionRecordTO.InstrumentRecord =
        CreateTransactionRecordTO.InstrumentRecord(
            accountId = uuidFrom(accountId),
            fundId = uuidFrom(fundId),
            amount = BigDecimal.parseString(amount),
            unit = Instrument(unitValue),
            labels = labels.map(::Label),
        )
}

File: client/web-client-react/src/jsMain/resources/react/api/transactionApi.ts
import { handleApiError } from './apiUtils';

export interface TransactionRecord {
    id: string;
    accountId: string;
    fundId: string;
    amount: string;
    unit: string;
    recordType: 'CURRENCY' | 'INSTRUMENT';
    labels: string[];
}

export interface Transaction {
    id: string;
    userId: string;
    externalId: string;
    dateTime: string;
    type: string;
    records: TransactionRecord[];
}

declare const window: Window & {
    FUNDS_CONFIG?: { fundServiceUrl?: string };
};

function getBaseUrl(): string {
    const url = window.FUNDS_CONFIG?.fundServiceUrl;
    if (!url) {
        throw new Error('FUNDS_CONFIG.fundServiceUrl is not configured');
    }
    return url;
}

const BASE_PATH = '/funds-api/fund/v1';

interface RawRecord {
    type: 'CURRENCY' | 'INSTRUMENT';
    id: string;
    accountId: string;
    fundId: string;
    amount: string;
    unit: string;
    labels: string[];
}

function mapRecord(raw: RawRecord): TransactionRecord {
    return {
        id: raw.id,
        accountId: raw.accountId,
        fundId: raw.fundId,
        amount: raw.amount,
        unit: raw.unit,
        recordType: raw.type,
        labels: raw.labels,
    };
}

function mapTransaction(raw: any): Transaction {
    const records: TransactionRecord[] = [];
    switch (raw.type) {
        case 'SINGLE_RECORD':
            records.push(mapRecord(raw.record));
            break;
        case 'TRANSFER':
            records.push(mapRecord(raw.sourceRecord));
            records.push(mapRecord(raw.destinationRecord));
            break;
        case 'EXCHANGE':
            records.push(mapRecord(raw.sourceRecord));
            records.push(mapRecord(raw.destinationRecord));
            if (raw.feeRecord) records.push(mapRecord(raw.feeRecord));
            break;
        case 'OPEN_POSITION':
        case 'CLOSE_POSITION':
            records.push(mapRecord(raw.currencyRecord));
            records.push(mapRecord(raw.instrumentRecord));
            break;
    }
    return {
        id: raw.id,
        userId: raw.userId,
        externalId: raw.externalId,
        dateTime: raw.dateTime,
        type: raw.type,
        records,
    };
}

export async function getTransaction(userId: string, transactionId: string): Promise<Transaction> {
    const url = `${getBaseUrl()}${BASE_PATH}/transactions/${transactionId}`;
    const response = await fetch(url, {
        headers: { 'FUNDS_USER_ID': userId },
    });
    if (!response.ok) await handleApiError(response, 'Failed to load transaction');
    const raw = await response.json();
    return mapTransaction(raw);
}
